merge form 1.10

Chelsea-Coder-Fan opened the pull request:
merge form 1.10

mdounin commented:
Please note that this repository is unofficial and pull requests have no chances to be merged. Proper way to submit changes to nginx is via the nginx development mailing list, see http://nginx.org/en/docs/contributing_changes.html.

PiotrSikora committed:
Configure: remove auto/lib/test, unused since nginx-0.1.2.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

arut committed:
Upstream: the "transparent" parameter of proxy_bind and friends.

This parameter lets binding the proxy connection to a non-local address.
Upstream will see the connection as coming from that address.
When used with $remote_addr, upstream will accept the connection from real
client address.

Example:

    proxy_bind $remote_addr transparent;

arut committed:
Upstream: prepared proxy_bind to accept parameters.

In addition, errors occurred while setting bind address are no longer ignored.

arut committed:
Stream: prepared proxy_bind to accept parameters.

mdocguard committed:
Version bump.

mdocguard committed:
Removed some bitrot.

Removed NGX_CONF_MULTI unused since 1.3.4.
Removed ngx_url_t.one_addr unused since 1.3.10.

vlhomutov committed:
Variable $request_id.

The variable contains text representation based on random data, usable as
a unique request identifier.

mdocguard committed:
Removed unused flag unexpected_eof from ngx_connection_t.

VBart committed:
Core: fixed port handling in ngx_parse_inet6_url().

This fixes buffer over-read when no port is specified in cases
similar to 5df5d7d771f6, and catches missing port separator.

mdocguard committed:
Removed a surplus condition from ngx_parse_inet6_url().

No functional changes, since the condition was always true.

xeioex committed:
Map: support of complex values in resulting strings.

VBart committed:
Improved EPOLLRDHUP handling.

When it's known that the kernel supports EPOLLRDHUP, there is no need in
additional recv() call to get EOF or error when the flag is absent in the
event generated by the kernel.  A special runtime test is done at startup
to detect if EPOLLRDHUP is actually supported by the kernel because
epoll_ctl() silently ignores unknown flags.

With this knowledge it's now possible to drop the "ready" flag for partial
read.  Previously, the "ready" flag was kept until the recv() returned EOF
or error.  In particular, this change allows the lingering close heuristics
(which relies on the "ready" flag state) to actually work on Linux, and not
wait for more data in most cases.

The "available" flag is now used in the read event with the semantics similar
to the corresponding counter in kqueue.

VBart committed:
Fixed an error log message about epoll_wait() timeout.

The errno value is unset in case of epoll_wait() timeout.

mdounin committed:
Dav: return 501 on PUT with ranges (ticket #948).

mdounin committed:
Cache: fixed updating bypassed cached errors (ticket #827).

mdounin committed:
Fixed work with --test-build-epoll after f7849bfb6d21.

mdounin committed:
Events: changed ngx_epoll_test_rdhup() to use existing epollfd.

mdounin committed:
Events: close descriptors on errors in ngx_epoll_test_rdhup().

mdounin committed:
Added overflow checks for version numbers (ticket #762).

Both minor and major versions are now limited to 999 maximum.  In case of
r->http_minor, this limit is already implied by the code.  Major version,
r->http_major, in theory can be up to 65535 with current code, but such
values are very unlikely to become real (and, additionally, such values
are not allowed by RFC 7230), so the same test was used for r->http_major.

arut committed:
Stream: return module.

mdounin committed:
SSL: support for multiple curves (ticket #885).

OpenSSL 1.0.2+ allows configuring a curve list instead of a single curve
previously supported.  This allows use of different curves depending on
what client supports (as available via the elliptic_curves extension),
and also allows use of different curves in an ECDHE key exchange and
in the ECDSA certificate.

The special value "auto" was introduced (now the default for ssl_ecdh_curve),
which means "use an internal list of curves as available in the OpenSSL
library used".  For versions prior to OpenSSL 1.0.2 it maps to "prime256v1"
as previously used.  The default in 1.0.2b+ prefers prime256v1 as well
(and X25519 in OpenSSL 1.1.0+).

As client vs. server preference of curves is controlled by the
same option as used for ciphers (SSL_OP_CIPHER_SERVER_PREFERENCE),
the ssl_prefer_server_ciphers directive now controls both.

mdounin committed:
OCSP stapling: staple now extracted via SSL_get_certificate().

This makes it possible to properly return OCSP staple with multiple
certificates configured.

Note that it only works properly in OpenSSL 1.0.1d+, 1.0.0k, 0.9.8y+.
In older versions SSL_get_certificate() fails to return correct certificate
when the certificate status callback is called.

mdounin committed:
OCSP stapling: additional function to configure stapling on a cert.

mdounin committed:
SSL: removed default DH parameters.

Using the same DH parameters on multiple servers is believed to be subject
to precomputation attacks, see http://weakdh.org/.  Additionally, 1024 bits
are not enough in the modern world as well.  Let users provide their own
DH parameters with the ssl_dhparam directive if they want to use EDH ciphers.

Note that SSL_CTX_set_dh_auto() as provided by OpenSSL 1.1.0 uses fixed
DH parameters from RFC 5114 and RFC 3526, and therefore subject to the same
precomputation attacks.  We avoid using it as well.

This change also fixes compilation with OpenSSL 1.1.0-pre5 (aka Beta 2),
as OpenSSL developers changed their policy after releasing Beta 1 and
broke API once again by making the DH struct opaque (see ticket #860).

mdounin committed:
SSL: style.

mdounin committed:
SSL: support for per-certificate chains.

The SSL_CTX_add0_chain_cert() function as introduced in OpenSSL 1.0.2 now
used instead of SSL_CTX_add_extra_chain_cert().

SSL_CTX_add_extra_chain_cert() adds extra certs for all certificates
in the context, while SSL_CTX_add0_chain_cert() only to a particular
certificate.  There is no difference unless multiple certificates are used,
though it is important when using multiple certificates.

Additionally, SSL_CTX_select_current_cert() is now called before using
a chain to make sure correct chain will be returned.

mdounin committed:
OCSP stapling: staple provided in arguments.

mdounin committed:
OCSP stapling: staple now stored in certificate, not SSL context.

mdounin committed:
SSL: support for multiple certificates (ticket #814).

mdounin committed:
SSL: error messages style.

mdounin committed:
SSL: made it possible to iterate though all certificates.

A pointer to a previously configured certificate now stored in a certificate.
This makes it possible to iterate though all certificates configured in
the SSL context.  This is now used to configure OCSP stapling for all
certificates, and in ngx_ssl_session_id_context().

As SSL_CTX_use_certificate() frees previously loaded certificate of the same
type, and we have no way to find out if it's the case, X509_free() calls
are now posponed till ngx_ssl_cleanup_ctx().

Note that in OpenSSL 1.0.2+ this can be done without storing things in exdata
using the SSL_CTX_set_current_cert() and SSL_CTX_get0_certificate() functions.
These are not yet available in all supported versions though, so it's easier
to continue to use exdata for now.

mdocguard committed:
Belatedly changed the ngx_create_listening() prototype.

The function is called only with "struct sockaddr *" since 0.7.58.

VBart committed:
HTTP/2: the "421 Misdirected Request" response (closes #848).

Since 4fbef397c753 nginx rejects with the 400 error any attempts of
requesting different host over the same connection, if the relevant
virtual server requires verification of a client certificate.

While requesting hosts other than negotiated isn't something legal
in HTTP/1.x, the HTTP/2 specification explicitly permits such requests
for connection reuse and has introduced a special response code 421.

According to RFC 7540 Section 9.1.2 this code can be sent by a server
that is not configured to produce responses for the combination of
scheme and authority that are included in the request URI.  And the
client may retry the request over a different connection.

Now this code is used for requests that aren't authorized in current
connection.  After receiving the 421 response a client will be able
to open a new connection, provide the required certificate and retry
the request.

Unfortunately, not all clients currently are able to handle it well.
Notably Chrome just shows an error, while at least the latest version
of Firefox retries the request over a new connection.

mdocguard committed:
Use ngx_cmp_sockaddr() where appropriate.

mdocguard committed:
Stream: fixed duplicate listen address detection.

The 6f8254ae61b8 change inadvertently fixed the duplicate port
detection similar to how it was fixed for mail in b2920b517490.
It also revealed another issue: the socket type (tcp vs. udp)
wasn't taken into account.

mdocguard committed:
Introduced the ngx_sockaddr_t type.

It's properly aligned and can hold any supported sockaddr.

mdounin committed:
Renamed "u" to "sockaddr" in listen options types.

xeioex committed:
Added the $proxy_protocol_port variable.

xeioex committed:
Realip: take client port from PROXY protocol header.

Previously, when the client address was changed to the one from
the PROXY protocol header, the client port ($remote_port) was
reset to zero.  Now the client port is also changed to the one
from the PROXY protocol header.

xeioex committed:
Added the $realip_remote_port variable.

xeioex committed:
Realip: port support in X-Real-IP and X-Forwarded-For.

Now, the module extracts optional port which may accompany an
IP address.  This custom extension is introduced, among other
things, in order to facilitate logging of original client ports.
Addresses with ports are expected to be in the RFC 3986 format,
that is, with IPv6 addresses in square brackets.  E.g.,
"X-Real-IP: [2001:0db8::1]:12345" sets client port ($remote_port)
to 12345.

mdocguard committed:
Realip: detect duplicate real_ip_header directive.

VBart committed:
HTTP/2: implemented preread buffer for request body (closes #959).

Previously, the stream's window was kept zero in order to prevent a client
from sending the request body before it was requested (see 887cca40ba6a for
details).  Until such initial window was acknowledged all requests with
data were rejected (see 0aa07850922f for details).

That approach revealed a number of problems:

 1. Some clients (notably MS IE/Edge, Safari, iOS applications) show an error
    or even crash if a stream is rejected;

 2. This requires at least one RTT for every request with body before the
    client receives window update and able to send data.

To overcome these problems the new directive "http2_body_preread_size" is
introduced.  It sets the initial window and configures a special per stream
preread buffer that is used to save all incoming data before the body is
requested and processed.

If the directive's value is lower than the default initial window (65535),
as previously, all streams with data will be rejected until the new window
is acknowledged.  Otherwise, no special processing is used and all requests
with data are welcome right from the connection start.

The default value is chosen to be 64k, which is bigger than the default
initial window.  Setting it to zero is fully complaint to the previous
behavior.

mdounin committed:
Fixed build on MSVC.

mdounin committed:
Updated OpenSSL used for win32 builds.

mdounin committed:
nginx-1.11.0-RELEASE

mdounin committed:
release-1.11.0 tag

VBart committed:
Version bump.

VBart committed:
HTTP/2: unbreak build on MSVC.

mdounin committed:
Trailing spaces removed.

mdounin committed:
Core: skip special buffers on writing (ticket #981).

A special last buffer with cl->buf->pos set to NULL can be present in
a chain when writing request body if chunked encoding was used.  This
resulted in a NULL pointer dereference if it happened to be the only
buffer left after a do...while loop iteration in ngx_write_chain_to_file().

The problem originally appeared in nginx 1.3.9 with chunked encoding
support.  Additionally, rev. 3832b608dc8d (nginx 1.9.13) changed the
minimum number of buffers to trigger this from IOV_MAX (typically 1024)
to NGX_IOVS_PREALLOCATE (typically 64).

Fix is to skip such buffers in ngx_chain_to_iovec(), much like it is
done in other places.

mdounin committed:
nginx-1.11.1-RELEASE

mdounin committed:
release-1.11.1 tag

pluknet committed:
Configure: revised GCC version processing.

Now GCC 6 and onwards will use -Wno-unused-parameter.

pluknet committed:
Version bump.

VBart committed:
Fixed an error log message.

ottok committed:
Fixed spelling.

mdounin committed:
An internal SHA1 implementation.

vlhomutov committed:
Stream: core module variables.

vlhomutov committed:
Stream: variables in proxy_pass and proxy_ssl_name.

vlhomutov committed:
Stream: added preconfiguration step.

ttaubert committed:
SSL: ngx_ssl_ciphers() to set list of ciphers.

This patch moves various OpenSSL-specific function calls into the
OpenSSL module and introduces ngx_ssl_ciphers() to make nginx more
crypto-library-agnostic.

VBart committed:
HTTP/2: fixed a segfault while processing unbuffered upload.

The ngx_http_v2_finalize_connection() closes current stream, but that is an
invalid operation while processing unbuffered upload.  This results in access
to already freed memory, since the upstream module sets a cleanup handler that
also finalizes the request.

VBart committed:
HTTP/2: fixed the "http request count is zero" alert.

When the stream is terminated the HEADERS frame can still wait in the output
queue.  This frame can't be removed and must be sent to the client anyway,
since HTTP/2 uses stateful compression for headers.  So in order to postpone
closing and freeing memory of such stream the special close stream handler
is set to the write event.  After the HEADERS frame is sent the write event
is called and the stream will be finally closed.

Some events like receiving a RST_STREAM can trigger the read handler of such
stream in closing state and cause unexpected processing that can result in
another attempt to finalize the request.  To prevent it the read handler is
now set to ngx_http_empty_handler.

Thanks to Amazon.

VBart committed:
HTTP/2: avoid adding Content-Length for requests without body.

There is no reason to add the "Content-Length: 0" header to a proxied request
without body if the header isn't presented in the original request.

Thanks to Amazon.

VBart committed:
HTTP/2: prevented double termination of a stream.

According to RFC 7540, an endpoint should not send more than one RST_STREAM
frame for any stream.

Also, now all the data frames will be skipped while termination.

defanator committed:
Set IP_BIND_ADDRESS_NO_PORT socket option for upstream sockets.

arut committed:
Introduced ngx_inet_get_port() and ngx_inet_set_port() functions.

arut committed:
Upstream: support for port in proxy_bind and friends.

arut committed:
Stream: support for $remote_port in proxy_bind.

The following two types of bind addresses are supported in addition to
$remote_addr and address literals:

- $remote_addr:$remote_port
- [$remote_addr]:$remote_port

In both cases client remote address with port is used in upstream socket bind.

arut committed:
Stream: set SO_REUSEADDR for UDP upstream sockets.

The option is only set if the socket is bound to a specific port to allow
several such sockets coexist at the same time.  This is required, for example,
when nginx acts as a transparent proxy and receives two datagrams from the same
client in a short time.

The feature is only implemented for Linux.

arut committed:
Fixed build on MSVC.

arut committed:
Stream: use ngx_pcalloc() in ngx_stream_proxy_bind().

arut committed:
Style.

PiotrSikora committed:
HTTP/2: style.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

arut committed:
Style.

PiotrSikora committed:
Configure: style.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

PiotrSikora committed:
Configure: fix build with -Werror=unused-value.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

PiotrSikora committed:
Configure: fix build with -Werror=unused-but-set-variable.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

PiotrSikora committed:
Configure: fix build with -Werror=nonnull.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

PiotrSikora committed:
Configure: fix build with -Werror=old-style-definition.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

PiotrSikora committed:
Configure: stop polluting NGX_ namespace.

While there, fix the only test that used alternative variable name.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

vlhomutov committed:
Stream: map module.

vlhomutov committed:
Stream: got rid of pseudo variables.

Stream limit_conn, upstream_hash and proxy modules now use complex values.

vlhomutov committed:
Stream: SSL-related variables.

mdocguard committed:
Removed unused flag accept_context_updated from ngx_event_t.

Also, removed practically unused flag accept_context_updated from
ngx_connection_t.

vlhomutov committed:
Stream: geo module.

mdounin committed:
Internal md5 and sha1 implementations are now always used.

This reduces the number of moving parts in ABI compatibility checks.
Additionally, it also allows to use OpenSSL in FIPS mode while still
using md5 for non-security tasks.

arut committed:
Sub filter: introduced the ngx_http_sub_match() function.

No functional changes.

arut committed:
Sub filter: eliminate unnecessary buffering.

Previously, when a buffer was processed by the sub filter, its final bytes
could be buffered by the filter even if they don't match any pattern.
This happened because the Boyer-Moore algorithm, employed by the sub filter
since b9447fc457b4 (1.9.4), matches the last characters of patterns prior to
checking other characters.  If the last character is out of scope, initial
bytes of a potential match are buffered until the last character is available.

Now, after receiving a flush or recycled buffer, the filter performs
additional checks to reduce the number of buffered bytes.  The potential match
is checked against the initial parts of all patterns.  Non-matching bytes are
not buffered.  This improves processing of a chunked response from upstream
by sending the entire chunks without buffering unless a partial match is found
at the end of a chunk.

vlhomutov committed:
Stream: variables and script.

This is a port of corresponding http code with unrelated features excluded.

mdounin committed:
Updated PCRE used for win32 builds.

mdounin committed:
release-1.11.2 tag

mdounin committed:
nginx-1.11.2-RELEASE

mdocguard committed:
Version bump.

mdocguard committed:
Use NGX_MAX_PATH_LEVEL where appropriate.

The macro was unused since 0.7.44.

vlhomutov committed:
Fixed regex captures handling without PCRE.

If PCRE is disabled, captures were treated as normal variables in
ngx_http_script_compile(), while code calculating flushes array length in
ngx_http_compile_complex_value() did not account captures as variables.
This could lead to write outside of the array boundary when setting
last element to -1.

Found with AddressSanitizer.

vlhomutov committed:
Stream: resolver.

pluknet committed:
Avoid left-shifting integers into the sign bit, which is undefined.

Found with UndefinedBehaviorSanitizer.

pluknet committed:
HTTP/2: avoid left-shifting signed integer into the sign bit.

On non-aligned platforms, properly cast argument before left-shifting it in
ngx_http_v2_parse_uint32 that is used with u_char.  Otherwise it propagates
to int to hold the value and can step over the sign bit.  Usually, on known
compilers, this results in negation.  Furthermore, a subsequent store into a
wider type, that is ngx_uint_t on 64-bit platforms, results in sign-extension.

In practice, this can be observed in debug log as a very large exclusive bit
value, when client sent PRIORITY frame with exclusive bit set:

: *14 http2 PRIORITY frame sid:5 on 1 excl:8589934591 weight:17

Found with UndefinedBehaviorSanitizer.

vlhomutov committed:
Stream: individual build options for modules.

vlhomutov committed:
Stream: style.

vlhomutov committed:
Stream: geoip module.

vlhomutov committed:
Stream: split_clients module.

VBart committed:
Events: the "accept_mutex" directive is turned off by default.

Now it is believed that the accept mutex brings more harm than benefits.
Especially in various benchmarks it often results in situation where only
one worker grabs all connections.

VBart committed:
Events: support for EPOLLEXCLUSIVE.

This flag appeared in Linux 4.5 and is useful for avoiding thundering herd
problem.

The current Linux kernel implementation walks the list of exclusive waiters,
and queues an event to each epfd, until it finds the first waiter that has
threads blocked on it via epoll_wait().

VBart committed:
Style: sorted epoll flags.

VBart committed:
HTTP/2: prevented output of the HEADERS frame for canceled streams.

It's useless to generate HEADERS if the stream has been canceled already.

VBart committed:
HTTP/2: always send GOAWAY while worker is shutting down.

Previously, if the worker process exited, GOAWAY was sent to connections in
idle state, but connections with active streams were closed without GOAWAY.

VBart committed:
HTTP/2: always handle streams in error state.

Previously, a stream could be closed by timeout if it was canceled
while its send window was exhausted.

VBart committed:
HTTP/2: avoid sending output queue if there's nothing to send.

Particularly this fixes alerts on OS X and NetBSD systems when HTTP/2 is
configured over plain TCP sockets.

On these systems calling writev() with no data leads to EINVAL errors
being logged as "writev() failed (22: Invalid argument) while processing
HTTP/2 connection".

VBart committed:
HTTP/2: fixed send timer handling.

Checking for return value of c->send_chain() isn't sufficient since there
are data can be left in the SSL buffer.  Now the wew->ready flag is used
instead.

In particular, this fixed a connection leak in cases when all streams were
closed, but there's still some data to be sent in the SSL buffer and the
client forgot about the connection.

VBart committed:
HTTP/2: refactored ngx_http_v2_send_output_queue().

Now it returns NGX_AGAIN if there's still data to be sent.

VBart committed:
HTTP/2: flushing of the SSL buffer in transition to the idle state.

It fixes potential connection leak if some unsent data was left in the SSL
buffer.  Particularly, that could happen when a client canceled the stream
after the HEADERS frame has already been created.  In this case no other
frames might be produced and the HEADERS frame alone didn't flush the buffer.

mdounin committed:
nginx-1.11.3-RELEASE

mdounin committed:
release-1.11.3 tag

vlhomutov committed:
Version bump.

vlhomutov committed:
Stream: fixed build without stream_ssl_module (ticket #1032).

mdocguard committed:
Win32: added per-thread random seeding.

The change in b91bcba29351 was not enough to fix random() seeding.
On Windows, the srand() seeds the PRNG only in the current thread,
and worse, is not inherited from the calling thread.  Due to this,
worker threads were not properly seeded.

Reported by Marc Bevand.

mdocguard committed:
Fixed undefined behavior when left shifting signed integer.

mdocguard committed:
Always seed PRNG with PID, seconds, and milliseconds.

pluknet committed:
SSL: guarded SSL_R_NO_CIPHERS_PASSED not present in OpenSSL 1.1.0.

It was removed in OpenSSL 1.1.0 Beta 3 (pre-release 6).  It was
not used since OpenSSL 1.0.1n and 1.0.2b.

mdocguard committed:
Simplified extraction of current time.

vlhomutov committed:
Upstream: the $upstream_bytes_received variable.

Unlike $upstream_response_length that only counts the body size,
the new variable also counts the size of response header and data
received after switching protocols when proxying WebSockets.

arut committed:
Ensure "listen" exists in a mail or stream server (ticket #1049).

VBart committed:
Geo: fixed access to already freed memory.

Previously, in "ranges" mode when all added ranges were deleted,
the ctx.high.low[i] was left pointing to a temporary array.

arut committed:
Fixed wrong type of the "line" field.

The new type ngx_uint_t was supposed when formatting the line number.

PiotrSikora committed:
SSL: remove no longer needed workaround for BoringSSL.

BoringSSL added a no-op stub for OPENSSL_config() on 2016-01-26.

Requested by David Benjamin.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

mdocguard committed:
Events: fixed setting of IP_BIND_ADDRESS_NO_PORT/SO_REUSEADDR.

The IP_BIND_ADDRESS_NO_PORT option is set on upstream sockets
if proxy_bind does not specify a port.  The SO_REUSEADDR option
is set on UDP upstream sockets if proxy_bind specifies a port.

Due to checking of the wrong port, IP_BIND_ADDRESS_NO_PORT was
never set, and SO_REUSEPORT was always set.

pluknet committed:
SSL: adopted session ticket handling for OpenSSL 1.1.0.

Return 1 in the SSL_CTX_set_tlsext_ticket_key_cb() callback function
to indicate that a new session ticket is created, as per documentation.
Until 1.1.0, OpenSSL didn't make a distinction between non-negative
return values.

See https://git.openssl.org/?p=openssl.git;a=commitdiff;h=5c753de for details.

mdocguard committed:
Geo: fixed overflow when iterating over ranges.

mdocguard committed:
Geo: fixed removing a range in certain cases.

If the range includes two or more /16 networks and does
not start at the /16 boundary, the last subrange was not
removed (see 91cff7f97a50 for details).

mdocguard committed:
Geo: fixed insertion of ranges specified in descending order.

mdocguard committed:
Geo: fixed warnings when removing nonexistent ranges.

geo $geo {
    ranges;

    10.0.0.0-10.0.0.255 test;

    delete 10.0.1.0-10.0.1.255;     # should warn

    delete 10.0.0.0-10.0.0.255;
    delete 10.0.0.0-10.0.0.255;     # should warn
}

mdounin committed:
Contrib: unicode2nginx compatibility with recent Perl versions.

In recent Perl versions unpack("C*") unpacks wide characters by default,
likely since perl 5.10 (seen at least in perl 5.20).  Replaced with
unpack("U0C*") instead to unpack bytes.

While here, improved style and updated my email.

pluknet committed:
Geo: fixed indentation.

thresheek commented:
Pull request on github are not accepted.  Please contribute via http://nginx.org/en/docs/contributing_changes.html. Thank you.
